lib/pos_network.go

```diff
@@ -384,19 +384,143 @@ func (sig *AggregatedBLSSignature) ToBytes() ([]byte, error) {
 	return retBytes, nil
 }
 
-func DecodeAggregatedBLSSignature(rr io.Reader) (*AggregatedBLSSignature, error) {
-	var sig AggregatedBLSSignature
+func (sig *AggregatedBLSSignature) FromBytes(rr io.Reader) error {
 	var err error
 
 	sig.SignersList, err = DecodeByteArray(rr)
 	if err != nil {
-		return nil, errors.Wrapf(err, "DecodeAggregatedBLSSignature: Error decoding SignersList")
+		return errors.Wrapf(err, "AggregatedBLSSignature.FromBytes: Error decoding SignersList")
 	}
 
 	sig.Signature, err = DecodeBLSSignature(rr)
 	if err != nil {
-		return nil, errors.Wrapf(err, "DecodeAggregatedBLSSignature: Error decoding Signature")
+		return errors.Wrapf(err, "AggregatedBLSSignature.FromBytes: Error decoding Signature")
+	}
+
+	return nil
+}
+
+// TimeoutAggregateQuorumCertificate is an aggregation of timeout messages from 2/3rds
+// of all validators, weighted by stake, which indicates that these validators want to
+// time out a particular view.
+//
+// When validators want to time out a view, they send their high QCs to the block proposer
+// who builds an aggregate QC extending the chain from the highest QC that it received
+// from all validators who timed out. To prove that it has selected the highest QC, the
+// proposer also includes a list of the high QC views that each validator has sent.
+type TimeoutAggregateQuorumCertificate struct {
+
+	// The view that the block proposers has produced a timeout QC for.
+	TimedOutView uint64
+
+	// This is the highest QC that the block proposer received from any validator who
+	// has timed out for the current view.
+	ValidatorsHighQC *QuorumCertificate
+
+	// Here we include a list of the HighQC.View values we got from each of the
+	// validators in the ValidatorsTimeoutHighQCViews field. In addition, for each
+	// unique HighQC.View value we received, we combine all the partial signatures
+	// for that HighQC.View into a single BLSMultiSignature.
+	//
+	// Note: The aggregated signature is made up of partial signatures for all
+	// present validators, each of whom signed a payload with the pair
+	// (current view, the validator's local HighQC.View).
+	ValidatorsTimeoutHighQCViews         []uint64
```

Comment:
How is this slice ordered? do we need to be concerned about non-determinism in how these uint64s are ordered? or does each index map to a validator somehow? If so, can we add a little bit of context about the ordering please?

Reply:
The ordering is decoupled from the schema here. I can add a comment here for context though.

The ordering of view numbers here will correspond to the ordering of the active validators in the `SnapshotValidatorByStatusAndStake` badger index. The index orders validators by `EpochNumber > Status > TotalStakeAmountNanos > ValidatorPKID`.

This array will have the highest stake active validator index 0, and list the remaining active validators in decreasing order of stake.